### ISSUE TITLE
Tilemap rendering incorrectly when positioned exactly on .5

### ISSUE THREAD

**Issue by blDraX** [godotengine/godot#76517]
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Windows 11

### Issue description

When using typical settings for obtaining pixel-perfect rendering (texture filter nearest, stretch mode viewport), tilemaps do not render correctly when they (or the camera) are exactly positioned on a half-pixel (e.g. 0.5). Looks like the left-most column is duplicated and the right-most column is cut off. Only tested for a horizontal .5 coordinate.

https://user-images.githubusercontent.com/51884888/234953179-581a33e9-73e3-4cc0-a6e6-10aa57c88686.mp4

### Steps to reproduce

Set up new project with a tilemap. Build a simple tilemap. Set texture filter to nearest and stretch mode viewport. Make resolution very small to make issue more apparent.

### Minimal reproduction project

[tilemap_rendering.zip](https://github.com/godotengine/godot/files/11346653/tilemap_rendering.zip)

**Comment by groud** [comment]
I checked the code, but could not find anything wrong on the TileMap side. I think there's likely an issue with the rendering code but I have no clue what would cause that.

As a workaround, it seems like disabling `Use texture padding` helps with it.

**Comment by Cerno-b** [comment]
Is it related to this? 82696

**Comment by adamscott** [comment]
I cannot reproduce, either on 4.0.4 or 4.3.dev, with the supplied MRP.

Can you confirm that the issue is still relevant, @blDraX?

**Comment by blDraX** [comment]
@adamscott: Yes, I can still reproduce this with the minimal example I've attached on v4.2.1.stable.official [b09f793f5].

In fact, while also testing this on my pixel art project, I found that the camera being close to .5 does not only do weird things to the tilemap but to my animated sprites as well. Check out this clip:
https://www.youtube.com/watch?v=J7paPFCq0H8

The diamonds moving up and down is not a change in position but implemented as animated sprite with three frames. Attaching the texture I'm using as well. When the camera is on .5, the pixels are doing.. super weird stuff? Attaching the png as well.

![diamond](https://github.com/godotengine/godot/assets/51884888/ac1aea0b-35ca-42cd-9b44-627b3a149f6b)

(in the video you can also see the problem on the tilemap again. Check out the side of the cliffs for example. Outmost pixel column is duplicated and on the other side, one column is cut off; also attaching a marked screenshot)
![tilemap_issue](https://github.com/godotengine/godot/assets/51884888/93de2851-71fd-4ef1-b64a-78bf338f8c33)

**Comment by adamscott** [comment]
@blDraX What should I see then in the MRP you joined?

**Comment by blDraX** [comment]
@adamscott: Sorry, I probably didn't choose a good texture in the MRP that easily shows the issue. But you should see essentially what I've also shown in the first video back in April last year. Change the camera position between 0 and 0.5 for example. You will see the tile texture wandering by one pixel column.

Here a recording using the MRP:
https://www.youtube.com/watch?v=aDwzO_K3U9I

**Comment by blDraX** [comment]
@adamscott @Cerno-b @groud I would like to bring attention to this issue again as it is still unsolved for me.

@groud previously mentioned that disabling texture padding can be used as a workaround. It's true that this prevents this specific rendering issue from happening but instead, a new issue arises: It seems like in the situations where the bug previously occured, the respective rows/columns are now instead not drawn at all:
![image](https://github.com/godotengine/godot/assets/51884888/90a099ce-4c9d-4e70-804f-8607ea5ba557)

It's hard for me to understand what is going on here and why only certain tiles are affected. I also have the issue attached on video which should make it a bit more clear what is going on.

I also want to add that this issue is apparently not limited to tilesets but also applies to ohter kinds of tiles images like for AnimatedSprite2D. In the second part of the video, you see the player sprite getting weirdly distorted as the camera moves diagonally, following the movement of the player up the slope:
![image](https://github.com/godotengine/godot/assets/51884888/1f9c9d9e-9ed6-4afc-9d50-466fcc2512fc)

For example, the lines appearing above the player sprite sometimes is actually the bottom row of one of the frames of the animation (the feet of the player sprite):
![image](https://github.com/godotengine/godot/assets/51884888/25ff1f5c-933a-4c8b-a909-2e7bc762d502)

For reference, this is a part of the respective sprite sheet I'm using (tiled 32x32):
![image](https://github.com/godotengine/godot/assets/51884888/0029619b-051f-4ecf-b9e8-c129c072200d)

Here is the video:
https://github.com/godotengine/godot/assets/51884888/c27fd49b-4c4a-4189-951d-88ed0437fdbb

**Comment by RandomGamingDev** [comment]
I can confirm that this issue is still present on the latest stable release ([4.3-stable](https://github.com/godotengine/godot/releases/tag/4.3-stable))

**Comment by RandomGamingDev** [comment]
I've published a PR (#97891) which fixes this issue whose scope seems to include things like Sprite2D outside of the original Tileset issue, meaning that it encompasses other issues outside of this one, although I wasn't able to check them all.

This fixes the original issue detailed, but would you be able to check which issues regarding pixel rendering were fixed for your project @blDraX?

**Comment by adamscott** [comment]
cc. @markdibarry

**Comment by markdibarry** [comment]
Hm. I get a segfault when trying to load the MRP, both in PR branch and in master. I'll tinker a bit.
